File: plugin/src/main/groovy/ru/cleverpumpkin/badge/ResourceUtils.java
package ru.cleverpumpkin.badge;

import org.xml.sax.SAXException;

import java.awt.Color;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import javax.xml.parsers.ParserConfigurationException;

import groovy.util.XmlSlurper;
import groovy.util.slurpersupport.GPathResult;

import static java.util.Collections.unmodifiableList;

public class ResourceUtils {

    private static XmlSlurper xmlSlurper;

    static {
        try {
            xmlSlurper = new XmlSlurper();
        } catch (ParserConfigurationException | SAXException e) {
            e.printStackTrace();
        }
    }

    public static String resourceFilePattern(String name) {
        if (name.startsWith("@")) {
            String[] pair = name.substring(1).split("/", 2);
            String baseResType = pair[0];
            String fileName = Objects.requireNonNull(pair[1], () ->
                    "Icon names does include resource types (e.g. drawable/ic_launcher):" + name
            );
            return baseResType + "*/" + fileName + ".*";
        }
        return name;

    }

    public static List<String> getLauncherIcons(File manifestFile) throws SAXException, IOException {

        GPathResult manifestXml = xmlSlurper.parse(manifestFile);
        GPathResult applicationNode = (GPathResult) manifestXml.getProperty("application");

        String icon = String.valueOf(applicationNode.getProperty("@android:icon"));
        String roundIcon = String.valueOf(applicationNode.getProperty("@android:roundIcon"));

        List<String> icons = new ArrayList<>(2);
        if (!icon.isEmpty()) icons.add(icon);
        if (!roundIcon.isEmpty()) icons.add(roundIcon);

        return unmodifiableList(icons);
    }

    /**
     * </p>Parse the color string, and return the corresponding color-int.
     * If the string cannot be parsed, throws an IllegalArgumentException
     * exception. Supported formats are:</p>
     *
     * <ul>
     * <li><code>#RRGGBB</code></li>
     * <li><code>#AARRGGBB</code></li>
     * </ul>
     */
    public static Color parseColor(String colorString) {
        if (colorString.charAt(0) == '#') {
            // Use a long to avoid rollovers on #ffXXXXXX
            long color = Long.parseLong(colorString.substring(1), 16);
            if (colorString.length() == 7) {
                // Set the alpha value
                color |= 0x00000000ff000000;
            } else if (colorString.length() != 9) {
                throw new IllegalArgumentException("Unknown color");
            }
            return new Color((int) color, true);
        }
        throw new IllegalArgumentException("Unknown color: " + colorString);
    }
}

File: plugin/src/main/java/ru/cleverpumpkin/appbadge/utils/ResourceUtils.kt
package ru.cleverpumpkin.appbadge.utils

import groovy.util.XmlSlurper
import groovy.util.slurpersupport.GPathResult
import org.xml.sax.SAXException
import java.awt.Color
import java.io.File
import java.io.IOException

/**
 * @author Sergey Chuprin
 */
object ResourceUtils {

    private const val MANIFEST_ICON_PROPERTY = "@android:icon"
    private const val MANIFEST_APPLICATION_PROPERTY = "application"
    private const val MANIFEST_ROUND_ICON_PROPERTY = "@android:roundIcon"

    private val xmlSlurper = XmlSlurper()

    @JvmStatic
    fun resourceFilePattern(name: String): String {
        if (!name.startsWith("@")) {
            return name
        }
        val expectedPair = name.substring(1).split("/".toRegex(), 2)
        val resType = expectedPair.first()
        val filename = requireNotNull(expectedPair.getOrNull(1)) {
            "Icon names does include resource types (e.g. drawable/ic_launcher):$name"
        }
        return "$resType*/$filename.*"
    }

    @JvmStatic
    @Throws(SAXException::class, IOException::class)
    fun getLauncherIcons(manifestFile: File): List<String> {
        val applicationNode = xmlSlurper
            .parse(manifestFile)
            .getProperty(MANIFEST_APPLICATION_PROPERTY) as GPathResult

        val icon = applicationNode.getProperty(MANIFEST_ICON_PROPERTY).toString()
        val roundIcon = applicationNode.getProperty(MANIFEST_ROUND_ICON_PROPERTY).toString()

        return listOfNotNull(icon.takeIf(String::isNotEmpty), roundIcon.takeIf(String::isNotEmpty))
    }

    /**
     * Parse the color string, and return the corresponding color-int.
     * If the string cannot be parsed, throws an IllegalArgumentException
     * exception. Supported formats are:
     * #RRGGBB
     * #AARRGGBB
     */
    @JvmStatic
    fun parseColor(colorString: String): Color {
        if (colorString[0] == '#') {

            // Use a long to avoid rollovers on #ffXXXXXX.
            val colorLong = colorString.substring(1).toLong(16)
            val resultColor = when {
                // Set alpha value.
                colorString.length == 7 -> colorLong or -0x1000000
                colorString.length != 9 -> {
                    throw IllegalArgumentException("Unknown color: $colorString")
                }
                else -> colorLong
            }
            return Color(resultColor.toInt(), true)
        }
        throw IllegalArgumentException("Unknown color: $colorString")
    }

}